# MSV / SS 2023 - Übung 10

## 10.1 Word2Vec Wiederholung

In [1]:
import gensim
%matplotlib inline

Daten: Google News corpus, 6B tokens (1M häufigsten Wörter)

Es ist wahrscheinlich, dass Sie eine erhöhte Datenrat brauchen (``jupyter notebook --NotebookApp.iopub_data_rate_limit=10000000`` im Terminal)

In [2]:
import gensim.downloader as api
wv = api.load('word2vec-google-news-300')

#### Embeddingsvektor für <i>King</i>

In [3]:
vec_king = wv['king']
vec_king

array([ 1.25976562e-01,  2.97851562e-02,  8.60595703e-03,  1.39648438e-01,
       -2.56347656e-02, -3.61328125e-02,  1.11816406e-01, -1.98242188e-01,
        5.12695312e-02,  3.63281250e-01, -2.42187500e-01, -3.02734375e-01,
       -1.77734375e-01, -2.49023438e-02, -1.67968750e-01, -1.69921875e-01,
        3.46679688e-02,  5.21850586e-03,  4.63867188e-02,  1.28906250e-01,
        1.36718750e-01,  1.12792969e-01,  5.95703125e-02,  1.36718750e-01,
        1.01074219e-01, -1.76757812e-01, -2.51953125e-01,  5.98144531e-02,
        3.41796875e-01, -3.11279297e-02,  1.04492188e-01,  6.17675781e-02,
        1.24511719e-01,  4.00390625e-01, -3.22265625e-01,  8.39843750e-02,
        3.90625000e-02,  5.85937500e-03,  7.03125000e-02,  1.72851562e-01,
        1.38671875e-01, -2.31445312e-01,  2.83203125e-01,  1.42578125e-01,
        3.41796875e-01, -2.39257812e-02, -1.09863281e-01,  3.32031250e-02,
       -5.46875000e-02,  1.53198242e-02, -1.62109375e-01,  1.58203125e-01,
       -2.59765625e-01,  

#### Wortähnlichkeit und Wortassoziationen

In [4]:
print(wv.most_similar(positive=['king'], topn=10))

[('kings', 0.7138046622276306), ('queen', 0.6510956883430481), ('monarch', 0.6413194537162781), ('crown_prince', 0.6204220056533813), ('prince', 0.6159993410110474), ('sultan', 0.5864824056625366), ('ruler', 0.5797566771507263), ('princes', 0.5646551847457886), ('Prince_Paras', 0.5432944297790527), ('throne', 0.5422104597091675)]


In [5]:
print(wv.most_similar(positive=['Hogwarts'], topn=10))

[('Hogwart', 0.7789777517318726), ('Death_Eaters', 0.6555673480033875), ('wizarding', 0.6494502425193787), ('Dumbledore', 0.6284919381141663), ('Harry_Daniel_Radcliffe', 0.6209906935691833), ('headmaster_Dumbledore', 0.6206334233283997), ('Hermione', 0.619870126247406), ('Voldemort', 0.6164700388908386), ('Harry_Potter', 0.6097619533538818), ('dementors', 0.6092057228088379)]


In [6]:
wv.similarity('Hogwarts', 'Dumbledore')

0.62849194

In [7]:
wv.similarity('Hogwarts', 'Sunnydale') # fictional setting for Buffy the Vampire Slayer 

0.29181698

In [8]:
wv.most_similar(positive=['king', 'woman'], negative=['man'], topn=1)

[('queen', 0.7118192911148071)]

In [9]:
wv.most_similar(positive=['because','teh'],negative=['the'])

[('becuase', 0.6815075278282166),
 ('becasue', 0.6744951605796814),
 ('cuz', 0.6165346503257751),
 ('becuz', 0.6027254462242126),
 ('coz', 0.580410361289978),
 ('b_c', 0.5737689733505249),
 ('tho', 0.5647958517074585),
 ('beacause', 0.5630674362182617),
 ('thats', 0.5605655908584595),
 ('lol', 0.5597798824310303)]

#### Analogien 

* 'Man' : 'King' = 'Woman': ?
* 'Man' verhält sich zu 'King' wie 'Woman' zu .....?

* vec("king") - vec("man") + vec("woman") =~ vec("queen")

In [10]:
wv.most_similar(positive=['king', 'woman'], negative=['man'], topn=1)

[('queen', 0.7118192911148071)]

In [11]:
wv.most_similar(positive=['sushi', 'Spain'], negative=['Japan'], topn=1)

[('tapas', 0.6775947213172913)]

In [12]:
wv.most_similar(positive=['computer_programmer', 'woman'], negative=['man'], topn=1)

[('homemaker', 0.5627119541168213)]

#### ... aber auch Bias!

In [13]:
wv.most_similar(positive=['screwdriver', 'she'], negative=['he'], topn=1)

[('knitting_needle', 0.5527439117431641)]

#### Crappie vs Guitar
    
![Image of a Crappie](https://upload.wikimedia.org/wikipedia/commons/thumb/4/4e/Pomoxis_nigromaculatus1.jpg/300px-Pomoxis_nigromaculatus1.jpg) 

In [14]:
wv.distance('crappie', 'guitar')

0.8140925616025925

#### "Dumb disambiguation" as vector sum with word2vec

In [15]:
print(wv.most_similar(positive=['bass','fish'], topn=10))

[('trout', 0.7888783812522888), ('striped_bass', 0.7623767256736755), ('largemouth_bass', 0.7479658126831055), ('crappie', 0.739653468132019), ('catfish', 0.7386457324028015), ('largemouth', 0.7284579873085022), ('smallmouth_bass', 0.7273966670036316), ('smallmouth', 0.7260484099388123), ('redfish', 0.7245889902114868), ('largemouths', 0.7244795560836792)]


In [16]:
print(wv.most_similar(positive=['bass','play'], topn=10))

[('playing', 0.6247290968894958), ('plays', 0.5736489295959473), ('played', 0.5574871897697449), ('acoustic_bass', 0.541383683681488), ('fingerstyle', 0.5367710590362549), ('topwater_bites', 0.5350249409675598), ('Spotted_bass', 0.5315738320350647), ('chordal_instrument', 0.5237420797348022), ('tenor_banjo', 0.5212451815605164), ('Niels_Henning_Orsted_Pedersen', 0.5189350247383118)]


### Word Embeddings mit SpaCy

In [17]:
import spacy
nlp_lg = spacy.load('en_core_web_lg')

In [18]:
text1 = "She learned to play the bass and joined a band"
text2 = "She went to her favorite fish restaurant and ordered the bass"

doc1_lg = nlp_lg(text1)
doc2_lg = nlp_lg(text2)

In [19]:
bass1_lg = doc1_lg[5]

In [20]:
bass2_lg = doc2_lg[10]

In [21]:
bass1_lg.vector[:30]

array([-1.9977 , -1.0515 ,  1.9286 ,  3.7206 , -0.16152, -1.2849 ,
       -2.2769 , -0.15887, -3.7156 , -0.44295,  1.0265 , -1.9462 ,
       -2.1289 , -0.74129,  4.0834 , -7.9771 ,  4.6809 ,  3.0406 ,
        8.7441 ,  2.2193 ,  1.8363 ,  0.02707,  6.9868 , -3.906  ,
        5.8547 ,  3.2925 , -5.5755 ,  0.22354,  4.2541 ,  0.50485],
      dtype=float32)

In [22]:
bass2_lg.vector[:30]

array([-1.9977 , -1.0515 ,  1.9286 ,  3.7206 , -0.16152, -1.2849 ,
       -2.2769 , -0.15887, -3.7156 , -0.44295,  1.0265 , -1.9462 ,
       -2.1289 , -0.74129,  4.0834 , -7.9771 ,  4.6809 ,  3.0406 ,
        8.7441 ,  2.2193 ,  1.8363 ,  0.02707,  6.9868 , -3.906  ,
        5.8547 ,  3.2925 , -5.5755 ,  0.22354,  4.2541 ,  0.50485],
      dtype=float32)

In [23]:
# Compare the similarity of the two meanings of 'bass'
bass1_lg.similarity(bass2_lg)

1.0

### 10.2 Contextuelle Embeddings

Quelle: https://applied-language-technology.mooc.fi/html/notebooks/part_iii/05_embeddings_continued.html

Sie sollen spacy-transformers installieren - ein Wrapper für Hugging Face’s transformers package

In [28]:
import spacy

nlp_trf = spacy.load("en_core_web_trf")

In [29]:
# Call the 'pipeline' attribute to examine the processing pipeline
# Am Anfang ist die Transformer-Component

nlp_trf.pipeline

[('transformer',
  <spacy_transformers.pipeline_component.Transformer at 0x7fe353e5c700>),
 ('tagger', <spacy.pipeline.tagger.Tagger at 0x7fe353e5c100>),
 ('parser', <spacy.pipeline.dep_parser.DependencyParser at 0x7fe323403370>),
 ('attribute_ruler',
  <spacy.pipeline.attributeruler.AttributeRuler at 0x7fe357e4cf00>),
 ('lemmatizer',
  <spacy.lang.en.lemmatizer.EnglishLemmatizer at 0x7fe350df4840>),
 ('ner', <spacy.pipeline.ner.EntityRecognizer at 0x7fe323403290>)]

In [30]:
text1 = "She learned to play the bass and joined a band"
text2 = "She went to her favorite fish restaurant and ordered the bass"

doc1_trf = nlp_trf(text1)
doc2_trf = nlp_trf(text2)

/Users/alessandra/anaconda3/lib/python3.10/site-packages/torch/amp/autocast_mode.py:198: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')


In [31]:
doc1_trf.__len__() # token length

10

In [32]:
doc2_trf.__len__() # token length

11

In [33]:
type(doc1_trf._.trf_data) # Transformer Data Object

spacy_transformers.data_classes.TransformerData

### Vortrainierte Räpresentationen

... als Tokensammlung:

In [34]:
doc1_trf._.trf_data.tensors[0].shape

(1, 12, 768)

eine "Batch" von 12 Vektoren (12 Tokens) mit jeweils 768 Dimensionen.

... als Sentence-Vektor:

In [35]:
doc1_trf._.trf_data.tensors[1].shape

(1, 768)

ein Vektor mit jeweils 768 Dimensionen.

Warum 12 Vektoren für 10 Tokens? Transformer-Modelle haben eine leicht unterschiedliche Tokenisierung

In [36]:
doc1_trf._.trf_data.tokens['input_texts']

[['<s>',
  'She',
  'Ġlearned',
  'Ġto',
  'Ġplay',
  'Ġthe',
  'Ġbass',
  'Ġand',
  'Ġjoined',
  'Ġa',
  'Ġband',
  '</s>']]

In [37]:
text3 = "Helsinki is the capital of Finland"
doc3_trf = nlp_trf(text3)

In [38]:
doc3_trf._.trf_data.tokens['input_texts'] 

[['<s>',
  'H',
  'els',
  'inki',
  'Ġis',
  'Ġthe',
  'Ġcapital',
  'Ġof',
  'ĠFinland',
  '</s>']]

nicht so häufige Wörter sind auf Sub-Wörten zerlegt, 'Ġ' wird zur Kennzeichnung von Leerzeichen verwendet

### Alignment zwischen Spacy-Tokens und Transformer-Tokens

Um die Spacy-Tokens und die Transformer-Tokens abzugleichen, verwenden wir das Attribut "align"

In [39]:
doc3_trf[0], doc3_trf._.trf_data.align[0].data

(Helsinki,
 array([[1],
        [2],
        [3]], dtype=int32))

In [40]:
doc3_trf[3], doc3_trf._.trf_data.align[3].data

(capital, array([[6]], dtype=int32))

#### Alignment für "bass" Sätze

In [41]:
doc1_trf

She learned to play the bass and joined a band

In [42]:
# bass als Instrument
doc1_trf[5]

bass

In [43]:
# Transformer Token Nr
n1 = doc1_trf._.trf_data.align[5].data[0][0]
n1

6

In [44]:
doc2_trf

She went to her favorite fish restaurant and ordered the bass

In [45]:
# bass als Fisch
doc2_trf[10]

bass

In [46]:
# Transformer Token Nr
n2 = doc2_trf._.trf_data.align[10].data[0][0]
n2

11

### Bass vs. bass

In [47]:
# [0]: es get um die einzelne Tokens
# [0]: diese Batch besteht aus 1 Vektor
# [n1]: token # n1
# [:30]: die erste 30 Dimensionen
doc1_trf._.trf_data.tensors[0][0][n1][:30]

array([ 0.26807055, -1.508554  , -0.60347795,  1.9065542 ,  2.605329  ,
        0.9219192 , -0.46964204,  0.5106118 , -0.19875261,  0.27527687,
       -1.0029974 , -1.8395708 , -0.34004122,  0.9450486 ,  1.1553266 ,
       -1.3235228 ,  2.2169526 , -1.012013  ,  1.1336058 ,  1.0907037 ,
       -0.44728377, -2.457325  ,  0.73260254,  0.00731617,  0.30130714,
        1.3202472 ,  0.39686862,  0.34239843,  1.2354615 , -1.5773274 ],
      dtype=float32)

In [48]:
bass1_trf = doc1_trf._.trf_data.tensors[0][0][n1]

In [49]:
doc2_trf._.trf_data.tensors[0][0][n2][:30]

array([ 0.48879564, -1.1985157 , -0.39528474,  1.2869403 ,  1.2567066 ,
        1.9768509 , -0.87547135,  0.6764189 , -0.02737204,  0.03502624,
       -1.3747454 , -1.7706739 , -0.36914095,  0.74596727,  0.46084306,
       -1.0485327 ,  2.596198  , -0.03518698,  1.084833  ,  1.2747976 ,
        0.12945312, -1.9378859 ,  0.6463141 ,  0.23293847,  0.70158815,
        1.4257596 ,  0.5985061 ,  0.32111108,  0.98047197, -0.53150326],
      dtype=float32)

In [50]:
bass2_trf = doc2_trf._.trf_data.tensors[0][0][n2]

In [51]:
from numpy import dot
from numpy.linalg import norm

cos_sim = dot(bass1_trf, bass2_trf)/(norm(bass1_trf)*norm(bass2_trf))
cos_sim

0.86093944

In [119]:
doc_fish_trf = nlp_trf("I was just done fileting my fish")
doc_guitar_trf = nlp_trf("I was just done tuning my guitar")

In [121]:
doc_fish_trf[6]

fish

In [125]:
doc_fish_trf._.trf_data.align[6].data[0]

array([8], dtype=int32)

In [122]:
doc_guitar_trf[6]

guitar

In [126]:
doc_guitar_trf._.trf_data.align[6].data[0]

array([7], dtype=int32)

In [127]:
fish_trf = doc_fish_trf._.trf_data.tensors[0][0][8]
guitar_trf = doc_guitar_trf._.trf_data.tensors[0][0][7]

#### Bass als Fisch vs. Guitar

In [128]:
cos_sim = dot(bass2_trf, guitar_trf)/(norm(bass2_trf)*norm(guitar_trf))
cos_sim

0.7972537

#### Bass als Instrument vs. Guitar

In [129]:
cos_sim = dot(bass1_trf, guitar_trf)/(norm(bass1_trf)*norm(guitar_trf))
cos_sim

0.80724585

#### Bass als Instrument vs. Fisch

In [130]:
cos_sim = dot(bass1_trf, fish_trf)/(norm(bass1_trf)*norm(fish_trf))
cos_sim

0.787453

#### Bass als Fisch vs. Crappie

In [131]:
cos_sim = dot(bass2_trf, fish_trf)/(norm(bass2_trf)*norm(fish_trf))
cos_sim

0.83117515

#### Crappie vs. Guitar

In [132]:
cos_sim = dot(crappie_trf, guitar_trf)/(norm(crappie_trf)*norm(guitar_trf))
cos_sim

0.53351605